src/array.c

```diff
@@ -215,10 +215,16 @@ JL_DLLEXPORT void jl_array_grow_end(jl_array_t *a, size_t inc)
     size_t oldmaxsize = a->ref.mem->length;
     size_t reqmaxsize = oldoffset + newnrows;
     if (__unlikely(reqmaxsize > oldmaxsize)) {
-        // grow to max(3/2oldmaxsize, reqmaxsize)
-        size_t newmaxsize = oldmaxsize*3/2;
+        size_t newmaxsize;
+        if (oldmaxsize < 4) // typical sequence: 0, // 4, // 6, 9, 13, 19, 28, 42, // 50, 60, 72, ...
+            newmaxsize = 4;
+        else if (oldmaxsize < 48)
+            newmaxsize = oldmaxsize*3/2; // grow by 50%
+        else
+            newmaxsize = oldmaxsize*6/5; // grow by 20%
         if (newmaxsize < reqmaxsize)
             newmaxsize = reqmaxsize;
+        // TODO: round this up to newmaxsize < GC_MAX_SZCLASS ? jl_gc_sizeclasses[jl_gc_szclass(newmaxsize)] : LLT_ALIGN(newmaxsize, 4096), after accounting for the object header (24 bytes)
```

Comment:
Am I understanding correctly that the growing heuristic is now different on the Julia side from the C API?

Reply:
yes

Reply:
the motivation is that the C resizing only happens for arrays in the compiler where we expect a lot of slow growing arrays, at which point saving memory matters more than saving time.